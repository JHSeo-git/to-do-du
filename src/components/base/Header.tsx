import React from "react";
import styled from "styled-components";
import { Link } from "react-router-dom";
import UserMenu from "components/base/UserMenu";
import UserMenuButton from "components/base/UserMenuButton";
import logo from "static/images/logo.svg";

const HeaderWrapper = styled.header`
  height: 4rem;
  padding: ${(props) => props.theme.space[5]};
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const HeaderColumn = styled.div`
  display: flex;
  align-items: center;
`;

const Logo = styled(Link)`
  margin-right: ${(props) => props.theme.space[3]};
`;

const LogoImg = styled.img`
  height: 40px;
`;

const MenuNav = styled.nav`
  display: flex;
  align-items: center;
`;

const MenuLink = styled(Link)`
  padding-top: ${(props) => props.theme.space[1]};
  padding-bottom: ${(props) => props.theme.space[1]};
  &:not(:last-child) {
    margin-right: ${(props) => props.theme.space[2]};
  }
`;

const Header = () => {
  return (
    <HeaderWrapper>
      <HeaderColumn>
        <Logo to="/">
          <LogoImg src={logo} alt="logo" />
        </Logo>
        <MenuNav>
          <MenuLink to="/">A Menu</MenuLink>
        </MenuNav>
      </HeaderColumn>
      <HeaderColumn>
        <UserMenuButton />
        <UserMenu />
      </HeaderColumn>
    </HeaderWrapper>
  );
};

export default Header;
