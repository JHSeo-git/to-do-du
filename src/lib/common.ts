export const pressedKey = (fn: () => void) => (e: React.KeyboardEvent) => {
  if (e.key === 'Enter') {
    fn();
  }

  return null;
};
