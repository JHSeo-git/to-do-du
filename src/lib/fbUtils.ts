import firebase from "firebase/app";
import { firebaseInstance } from "fBase";
import { cancelled, take, put } from "redux-saga/effects";
import { EventChannel, eventChannel, buffers } from "redux-saga";

export const getStringToAuthProvider = (provider: string) => {
  let authProvider: firebase.auth.AuthProvider | null = null;
  if (provider === "Github") {
    authProvider = new firebaseInstance.auth.GithubAuthProvider();
  } else if (provider === "Google") {
    authProvider = new firebaseInstance.auth.GoogleAuthProvider();
  } else if (provider === "Facebook") {
    authProvider = new firebaseInstance.auth.FacebookAuthProvider();
  }
  return authProvider;
};

export function makeChannel(
  collection: firebase.firestore.CollectionReference,
  buffer = buffers.none(),
  snapshotListenOptions: firebase.firestore.SnapshotListenOptions | undefined
) {
  const ref = collection;

  const channel = eventChannel((emit) => {
    const unsubscribe = snapshotListenOptions
      ? ref.onSnapshot(snapshotListenOptions, emit)
      : ref.onSnapshot(emit);

    // Returns unsubscribe function
    return unsubscribe;
  }, buffer);

  return channel;
}

// TODO: Refactoring
// export interface SyncChannelOptions {
//   onSuccess: typeof asyncSyncTodos.success;
//   onFailure: typeof asyncSyncTodos.failure;
//   snapshotListenOptions: firebase.firestore.SnapshotListenOptions | undefined;
//   transform: any;
// }

export function* syncChannel(channel: EventChannel<any>, options: any) {
  const { onSuccess, onFailure, transform } = options;

  try {
    while (true) {
      const data = yield take(channel);
      const transformedData = transform ? transform(data) : data;
      yield put(onSuccess(transformedData));
    }
  } catch (err) {
    if (onFailure) yield put(onFailure(err.message));
    else
      console.error(
        "The following error has been ignored because no `failureActionCreator` has been set:",
        err
      );
  } finally {
    if (yield cancelled()) {
      channel.close();
    }
  }
}
