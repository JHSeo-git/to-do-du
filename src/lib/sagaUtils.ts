import { AsyncActionCreatorBuilder, createAsyncAction } from 'typesafe-actions';
import { put, call, Effect, SimpleEffect } from 'redux-saga/effects';

type StripEffects<T> = T extends IterableIterator<infer E>
  ? E extends Effect | SimpleEffect<any, any>
    ? never
    : E
  : never;

type DecideReturn<T> = T extends Promise<infer R>
  ? R // If it's a promise, return the promised type.
  : T extends IterableIterator<any>
  ? StripEffects<T> // If it's a generator, strip any effects to get the return type.
  : T; // Otherwise, it's a normal function and the return type is unaffected.

export type CallReturnType<T extends (...args: any[]) => any> = DecideReturn<
  ReturnType<T>
>;

interface AsyncAction {
  REQUEST: string;
  SUCCESS: string;
  FAILURE: string;
}

export const asyncActionCreator = (actionName: string): AsyncAction => {
  const asyncTypeAction: string[] = ['_REQUEST', '_SUCCESS', '_FAILURE'];

  return {
    REQUEST: actionName + asyncTypeAction[0],
    SUCCESS: actionName + asyncTypeAction[1],
    FAILURE: actionName + asyncTypeAction[2],
  };
};

export const asyncAction = <T, P, J>(asyncAction: AsyncAction) => {
  return createAsyncAction(
    asyncAction.REQUEST,
    asyncAction.SUCCESS,
    asyncAction.FAILURE
  )<T, P, J>();
};

type PromiseCreatorFunction<P, T> =
  | ((payload: P) => Promise<T>)
  | (() => Promise<T>);

export default function createAsyncSaga<
  RequestType,
  RequestPayload,
  SuccessType,
  SuccessPayload,
  FailureType,
  FailurePayload
>(
  asyncAction: AsyncActionCreatorBuilder<
    [RequestType, [RequestPayload, undefined]],
    [SuccessType, [SuccessPayload, undefined]],
    [FailureType, [FailurePayload, undefined]]
  >,
  asyncFunction: PromiseCreatorFunction<RequestPayload, SuccessPayload>,
  successFunc?: any,
  failureFunc?: any
) {
  return function* saga(action: ReturnType<typeof asyncAction.request>) {
    try {
      const result: SuccessPayload = yield call(
        asyncFunction,
        (action as any).payload
      );
      yield put(asyncAction.success(result));
      if (successFunc) {
        yield call(successFunc, result);
      }
    } catch (e) {
      yield put(asyncAction.failure(e));
      if (failureFunc) {
        yield call(failureFunc, e);
      }
    }
  };
}
