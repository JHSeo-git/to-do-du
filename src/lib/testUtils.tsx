import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';
import { ThemeProvider } from 'styled-components';
import HelmetGlobal from 'components/base/HelmetGlobal';
import store from 'store';
import { GlobalStyle } from 'styles/globalStyles';
import { theme } from 'styles/theme';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Router } from 'react-router-dom';

interface Props {
  children: React.ReactNode;
}

interface RouteProps {
  route?: string;
  history?: MemoryHistory;
}

const AllTheProviders = ({ children }: Props) => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme.lightTheme}>
        <HelmetProvider>
          <HelmetGlobal />
          <GlobalStyle />
          {children}
        </HelmetProvider>
      </ThemeProvider>
    </Provider>
  );
};

const customRender = (ui: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const withRouterRender = (
  ui: any,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: RouteProps = {},
  options?: any
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>, {
      wrapper: AllTheProviders,
      ...options,
    }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
export { withRouterRender as renderWithRouter };
