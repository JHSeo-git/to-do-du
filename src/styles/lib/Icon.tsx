import styled from "styled-components";
import { lighten, darken } from "polished";
import { FaRegCircle, FaPlus, FaMinusCircle } from "react-icons/fa";
import { BiArrowFromLeft } from "react-icons/bi";

export const PlusIcon = styled(FaPlus)`
  color: ${(props) => props.theme.primaryColor};
`;

export const RegIcon = styled(FaRegCircle)`
  color: ${(props) => props.theme.grayDarkColor};
`;

export const DeleteIcon = styled(FaMinusCircle)`
  color: ${(props) => props.theme.alertColor};
  cursor: pointer;
  transition: all 0.2s linear;
  &:hover {
    color: ${(props) => lighten(0.1, props.theme.alertColor)};
  }
  &:active {
    color: ${(props) => darken(0.1, props.theme.alertColor)};
  }
`;

export const CloseIcon = styled(BiArrowFromLeft)`
  color: ${(props) => props.theme.grayDarkColor};
  cursor: pointer;
  transition: all 0.2s linear;
  &:hover {
    color: ${(props) => lighten(0.2, props.theme.grayDarkColor)};
  }
  &:active {
    color: ${(props) => darken(0.2, props.theme.grayDarkColor)};
  }
`;
