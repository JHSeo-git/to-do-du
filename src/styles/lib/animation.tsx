import { css, keyframes } from 'styled-components';

const fadeInKeyframes = keyframes`
    0%{
        opacity: 0;
    }
    100%{
        opacity: 1;
    }
`;

export const fadeInWithDelay = (delay = 0) => css`
  animation: ${fadeInKeyframes} 0.2s linear forwards;
  animation-delay: ${delay};
`;
