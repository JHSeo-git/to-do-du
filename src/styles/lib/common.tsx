import { css } from "styled-components";

export const whiteBox = css`
  border-radius: 4px;
  background: ${(props) => props.theme.whiteColor};
`;
