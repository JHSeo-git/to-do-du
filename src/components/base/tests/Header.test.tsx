import React from 'react';
import Header from 'components/base/Header';
import { renderWithRouter } from 'lib/testUtils';

describe('<Header />', () => {
  const renderFunc = () => {
    const utils = renderWithRouter(<Header />);
    return { ...utils };
  };

  it('matches snapshot', () => {
    const utils = renderFunc();
    expect(utils.container).toMatchSnapshot();
  });
});
