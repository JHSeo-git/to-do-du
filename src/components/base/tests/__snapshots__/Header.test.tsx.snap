// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> matches snapshot 1`] = `
<div>
  <header
    class="sc-jrAGrp hkJgOA"
  >
    <div
      class="sc-kEjbxe gXsult"
    >
      <a
        class="sc-iqHYGH dvthwX"
        href="/"
      >
        <img
          alt="logo"
          class="sc-crrsfI eJFwuc"
          src="logo.svg"
        />
      </a>
      <nav
        class="sc-dQppl QFlJH"
      >
        <a
          class="sc-bqyKva gspMji"
          href="/"
        />
      </nav>
    </div>
    <div
      class="sc-kEjbxe gXsult"
    >
      <div
        class="sc-gKsewC qQBTM"
      />
    </div>
  </header>
</div>
`;
